### PR TITLE
[rom_ext] fix certificate validity check

### LINKED ISSUE #22921: [rom_ext] ROM_EXT crashes during CDI1 serial number check
### Description

During regular FPGA test configuration, the device retains info flash page configuration between test cases. Running the following test cases result in ROM_EXT fault:

```shell
//sw/device/silicon_creator/lib/drivers:rstmgr_functest_fpga_cw310_sival_rom_ext
//sw/device/silicon_creator/lib/drivers:uart_functest_fpga_cw310_sival_rom_ext
```

Error signature

```log
Starting ROM_EXT
Warning: UDS certificate not valid.
BFV:02524902[2024-05-02T02:37:16Z INFO  opentitantool::command::console] ExitFailure("BFV:02524902")
```

Suspected function call triggering fault:

https://github.com/lowRISC/opentitan/blob/605da270ddcf70b2af5573e9524f9613bc89383d/sw/device/silicon_creator/rom_ext/rom_ext.c#L352-L354

Additional observations: The `rstmgr_functest_fpga_cw310_sival_rom_ext` test does not seem to be able to cache `CDI1`. Not clear why this is the case:

```log
Invoking test: sw/host/opentitantool/opentitantool --rcfile= --logging=info --interface=hyper310 --exec=transport init --exec=fpga load-bitstream hw/bitstream/universal/splice.bit --exec=bootstrap --clear-uart=true sw/device/silicon_creator/lib/drivers/rstmgr_functest_fpga_cw310_sival_rom_ext.img --exec=console --non-interactive --exit-success='PASS.*\n' --exit-failure='((FAIL|FAULT).*\n)|(BFV:[0-9a-f]{8})' no-op
[2024-05-02T02:36:53Z INFO  opentitantool] Command result: success.
[2024-05-02T02:36:53Z INFO  opentitantool::command::load_bitstream] Loading bitstream: "hw/bitstream/universal/splice.bit"
[2024-05-02T02:36:54Z INFO  opentitanlib::util::usr_access] Bitstream file USR_ACCESS value: 0xdd700156
[2024-05-02T02:36:54Z INFO  opentitanlib::util::rom_detect] Current bitstream: "ROM:dd700156\r"
[2024-05-02T02:36:54Z INFO  opentitanlib::transport::common::fpga] Already running the correct bitstream.  Skip loading bitstream.
[2024-05-02T02:36:54Z INFO  opentitantool] Command result: success.
[2024-05-02T02:36:54Z INFO  opentitanlib::bootstrap] Asserting bootstrap pins...
[2024-05-02T02:36:54Z INFO  opentitanlib::app] Asserting the reset signal
[2024-05-02T02:36:54Z INFO  opentitanlib::app] Clearing the UART RX buffer
[2024-05-02T02:36:54Z INFO  opentitanlib::app] Deasserting the reset signal
[2024-05-02T02:36:54Z INFO  opentitanlib::bootstrap] Performing bootstrap...
[2024-05-02T02:36:55Z INFO  opentitanlib::bootstrap] Releasing bootstrap pins, resetting device...
[2024-05-02T02:36:55Z INFO  opentitanlib::app] Asserting the reset signal
[2024-05-02T02:36:55Z INFO  opentitanlib::app] Deasserting the reset signal
[2024-05-02T02:36:55Z INFO  opentitantool] Command result: success.
[2024-05-02T02:36:55Z INFO  opentitanlib::io::uart] set_flow_control to false
ROM:dd700156
bootstrap:1
ROM:dd700156
ROM:dd700156
Starting ROM_EXT
Warning: UDS certificate not valid.
CDI_1 certificate not valid. Updating it ...
0: 40130000 NAPOT L--- sz=00001000
1: 40480000 NAPOT L--- sz=00000400
2: 20010400 ----- ---- sz=00000000
3: 20013904   TOR -X-R sz=00003504
4: 00000000 ----- ---- sz=00000000
5: 00000000 ----- ---- sz=00000000
6: 00000000 ----- ---- sz=00000000
7: 00000000 ----- ---- sz=00000000
8: 00000000 ----- ---- sz=00000000
9: 00000000 ----- ---- sz=00000000
10: 20000400 ----- ---- sz=00000000
11: 2000ae80   TOR -X-R sz=0000aa80
12: 20000000 NAPOT L--R sz=00100000
13: 00010000 NAPOT L--- sz=00001000
14: 40000000 NAPOT L-WR sz=10000000
15: 10000000 NAPOT L-WR sz=00020000
mseccfg = 00000002
entry: 0x20010480
I00001 ottf_main.c:154] Running sw/device/silicon_creator/lib/drivers/rstmgr_functest.c
I00002 rstmgr_functest.c:31] Reset reason: 0x00000001
ROM:dd700156
Starting ROM_EXT
Warning: UDS certificate not valid.
CDI_1 certificate not valid. Updating it ...
0: 40130000 NAPOT L--- sz=00001000
1: 40480000 NAPOT L--- sz=00000400
2: 20010400 ----- ---- sz=00000000
3: 20013904   TOR -X-R sz=00003504
4: 00000000 ----- ---- sz=00000000
5: 00000000 ----- ---- sz=00000000
6: 00000000 ----- ---- sz=00000000
7: 00000000 ----- ---- sz=00000000
8: 00000000 ----- ---- sz=00000000
9: 00000000 ----- ---- sz=00000000
10: 20000400 ----- ---- sz=00000000
11: 2000ae80   TOR -X-R sz=0000aa80
12: 20000000 NAPOT L--R sz=00100000
13: 00010000 NAPOT L--- sz=00001000
14: 40000000 NAPOT L-WR sz=10000000
15: 10000000 NAPOT L-WR sz=00020000
mseccfg = 00000002
entry: 0x20010480
I00001 ottf_main.c:154] Running sw/device/silicon_creator/lib/drivers/rstmgr_functest.c
I00002 rstmgr_functest.c:31] Reset reason: 0x00000004
I00003 rstmgr_functest.c:47] Detected software reset.
I00004 ottf_main.c:109] Finished sw/device/silicon_creator/lib/drivers/rstmgr_functest.c
I00005 status.c:28] PASS!
[2024-05-02T02:36:56Z INFO  opentitantool::command::console] ExitSuccess("PASS!\r\n")
[2024-05-02T02:36:56Z INFO  opentitantool] Command result: success.
[2024-05-02T02:36:56Z INFO  opentitantool] Command result: success.
cleanup:  
INFO: Elapsed time: 4.370s, Critical Path: 3.08s
INFO: 3 processes: 1 disk cache hit, 2 local.
INFO: Build completed successfully, 3 total actions
//sw/device/silicon_creator/lib/drivers:rstmgr_functest_fpga_cw310_sival_rom_ext PASSED in 2.9s
```

CC: @timothytrippel @pamaury

### PR DESCRIPTION
**Pull request by timothytrippel**
This fixes #22921, including errors in the byte ordering of certificate serial numbers, and how they are checked during ROM_EXT execution to detemine if the certificate needs to be updated.

Additionally, this adds some unit tests to test corner cases in the certificate serial number validity checking mechanism.

### CODE REVIEW

File: sw/device/silicon_creator/lib/cert/cert.c [hunk 1]
```diff
@@ -52,38 +54,45 @@ rom_error_t cert_x509_asn1_check_serial_number(
...
-    }
-    if (launder32(curr_sn_word) != expected_sn_words[i]) {
-      HARDENED_CHECK_NE(curr_sn_word, expected_sn_words[i]);
+
+  // Extract serial number bytes from the certificate into a 20-byte array as
+  // this is the full size of the serial number integer.
+  //
+  // We fill the array in reverse (starting at the last index, moving towards
+  // the first index) to ensure we maintain any zero padding.
+  for (size_t i = sn_bytes_offset + asn1_integer_length - 1,
```
Comment (pamaury) [inline sw/device/silicon_creator/lib/cert/cert.c:86]: Minor detail: this function does not clear the rest of `actual_serial_number` with zeroes and relies on it being zero before the call. Maybe it would be better if the function clear the buffer? Also, can this loop be replaced with a `memcpy`?

File: sw/device/silicon_creator/lib/cert/cert_unittest.cc [hunk 2]
```diff
@@ -200,14 +206,31 @@ TEST_F(CertTest, CertOutdated) {
...
+                          valid_dice_cert_words_, kErrorOk);
+  EXPECT_EQ(
+      cert_x509_asn1_check_serial_number(&kFlashCtrlInfoPageUdsCertificate,
+                                         expected_sn_bytes_, &matches),
+      kErrorOk);
+  EXPECT_EQ(matches, kHardenedBoolFalse);
+  valid_dice_cert_bytes_[kCertX509Asn1SerialNumberLengthByteOffset] =
+      old_length;
+}
+
```
Comment (pamaury) [inline sw/device/silicon_creator/lib/cert/cert_unittest.cc:225]: Should there be a test with a short serial number (19 bytes) that matches? Here the unit test don't really cover that as far as I can see.

**Review by pamaury — APPROVED** [review]
Thanks Tim, just a small question (see comment).

File: sw/device/silicon_creator/lib/cert/cert_unittest.cc (hunk 2 above)
Comment (timothytrippel) [reply sw/device/silicon_creator/lib/cert/cert_unittest.cc:225]: Ah, yes great idea thanks. I added one.

File: sw/device/silicon_creator/lib/cert/cert.c [hunk 3]
```diff
@@ -52,38 +54,52 @@ rom_error_t cert_x509_asn1_check_serial_number(
...
+  // larger serial number than currently present.
+  memset(actual_serial_number, 0, kCertX509Asn1SerialNumberSizeInBytes);
+  memcpy(&actual_serial_number[kCertX509Asn1SerialNumberSizeInBytes -
+                               asn1_integer_length],
+         &cert_bytes[sn_bytes_offset], asn1_integer_length);
+  /*for (size_t i = sn_bytes_offset + asn1_integer_length - 1,*/
+  /*j = kCertX509Asn1SerialNumberSizeInBytes - 1;*/
+  /*i >= sn_bytes_offset; --i, --j) {*/
+  /*actual_serial_number[j] = cert_bytes[i];*/
+  /*}*/
```
Comment (moidx) [inline sw/device/silicon_creator/lib/cert/cert.c:97]: nit: remove deadcode.

File: sw/device/silicon_creator/lib/cert/cert.c [hunk 4]
```diff
@@ -52,38 +54,52 @@ rom_error_t cert_x509_asn1_check_serial_number(
...
+
+  // Check the serial number in the certificate matches what was expected.
+  for (size_t i = 0; i < kCertX509Asn1SerialNumberSizeInBytes; ++i) {
+    if (launder32(actual_serial_number[i]) != expected_sn_bytes[i]) {
+      HARDENED_CHECK_NE(actual_serial_number[i], expected_sn_bytes[i]);
```
Comment (moidx) [inline sw/device/silicon_creator/lib/cert/cert.c:102]: nit: Since this function is pretty long, it may be worth adding `*matches = kHardenedBoolFalse` here.

Feel free to discard this comment on the assumption that this is already covered by tests.

File: sw/host/provisioning/ft_lib/src/lib.rs [hunk 5]
```diff
@@ -313,6 +313,14 @@ pub fn run_ft_personalize(
         log::info!("Running hash {:?}", hex::encode(hasher.clone().finalize()));
     }
 
+    let words = uds_cert_bytes
```
Comment (moidx) [inline sw/host/provisioning/ft_lib/src/lib.rs:316]: nit: Can you add the same comment you added on the C side describing why this is needed?

**Review by moidx — APPROVED** [review]

File: sw/device/silicon_creator/lib/cert/cert.c (hunk 4 above)
Comment (timothytrippel) [reply sw/device/silicon_creator/lib/cert/cert.c:98]: It is also covered by unit tests, but good idea, fixed.

File: sw/device/silicon_creator/lib/cert/cert.c (hunk 3 above)
Comment (timothytrippel) [reply sw/device/silicon_creator/lib/cert/cert.c]: oops. missed that. Fixed.

File: sw/host/provisioning/ft_lib/src/lib.rs (hunk 5 above)
Comment (timothytrippel) [reply sw/host/provisioning/ft_lib/src/lib.rs]: Oops, this was also debugging code I forgot to removed. Fixed. Thanks.
